### DISCUSSION TITLE
How to handle ‘RegisterFmpInstaller’ in FmpDxe to operate USBIO when updating capsule?

### DISCUSSION THREAD

**Discussion by gordontcp** [tianocore/edk2#3119] [Q&A]
Hi all,
In `\FmpDevicePkg\Library\FmpDeviceLibNull\FmpDeviceLib.c`
There are comments for `RegisterFmpInstaller`:
```
/**
  Used to pass the FMP install function to this lib.  This allows the library to
  have control of the handle that the FMP instance is installed on.  This allows
  the library to use DriverBinding protocol model to locate its device(s) in the
  system.

  @param[in] Func  Function pointer to FMP install function.

  @retval EFI_SUCCESS       Library has saved function pointer and will call
                            function pointer on each DriverBinding Start.
  @retval EFI_UNSUPPORTED   Library doesn't use driver binding and only supports
                            a single instance.
  @retval other error       Error occurred.  Don't install FMP

**/
EFI_STATUS
EFIAPI
RegisterFmpInstaller (
  IN FMP_DEVICE_LIB_REGISTER_FMP_INSTALLER Func
  )
{
  //
  // This is a system firmware update that does not use Driver Binding Protocol
  //
  return EFI_UNSUPPORTED;
}
```
However, since the ‘RegisterFmpInstaller’ of the example directly returns `EFI_UNSUPPORTED`, it does not demonstrate how to handle the DriverBinding protocol. There are similar instructions in the edk2-platforms example, but it also does not demonstrate how to locate the DriverBinding protocol. The full path of edk2-platforms example is :
`edk2-platforms\Platform\Intel\Vlv2TbltDevicePkg\Feature\Capsule\Library\FmpDeviceLib\FmpDeviceLib.c`

I would like to operate USBIO in FmpDxe, how to do Binding protocol in FmpDxe, is there any sample code?
Any suggestion is greatly appreciated!
Many thanks!

**Comment by spbrogan** [comment]
I have written a couple of instances that used driver binding and the above function.  In those implementations, I captured the function pointer into a global variable and then returned EFI_SUCCESS.  Then later when driver binding start runs your code can call the cached function pointer with the handle you now have.   I thought @mdkinney authored an open-source sample driver for an option rom that supported multi-instance and used this function.  I can't find it but maybe he will see this and respond.

**Reply by gordontcp** [reply]
Hi,
    I used your method :
`captured the function pointer into a global variable and then returned EFI_SUCCESS. Then later when driver binding start runs your code can call the cached function pointer with the handle now have. `
  However, I got following error messages:
```
 FmpDxe(WDT Firmware Device): Failed to lock variable DC1F7882-4C84-4F8B-9F0A-AC5301DD6FB0 FmpVersion. Status = Write Protected                            
 FmpDxe(WDT Firmware Device): Failed to lock variable DC1F7882-4C84-4F8B-9F0A-AC5301DD6FB0 FmpLsv. Status = Write Protected                                
 FmpDxe(WDT Firmware Device): Failed to lock variable DC1F7882-4C84-4F8B-9F0A-AC5301DD6FB0 LastAttemptStatus. Status = Write Protected                     
 FmpDxe(WDT Firmware Device): Failed to lock variable DC1F7882-4C84-4F8B-9F0A-AC5301DD6FB0 LastAttemptVersion. Status = Write Protected                    
 FmpDxe(WDT Firmware Device): Failed to lock variable DC1F7882-4C84-4F8B-9F0A-AC5301DD6FB0 FmpState. Status = Write Protected                              
 FmpDxe(WDT Firmware Device): Failed to register variables to lock.  Status = Write Protected.                                                             

```
```
ProcessFmpCapsuleImage - Not Ready
ProcessThisCapsuleImage - Not Ready
```
   Therefore, I can't finish capsule update image.
   Is there any example to demostrate the implementation? Any suggestion is highly appreciated.
Thanks!
